File: app/src/main/java/com/weathercock/android_cleanarchitecture/data/repository/WeatherRepositoryImpl.java
package com.weathercock.android_cleanarchitecture.data.repository;

import android.content.Context;

import com.weathercock.android_cleanarchitecture.data.api.WeatherApi;
import com.weathercock.android_cleanarchitecture.data.api.WeatherRetrofit;
import com.weathercock.android_cleanarchitecture.data.dao.WeatherDao;
import com.weathercock.android_cleanarchitecture.data.entity.WeatherEntity;
import com.weathercock.android_cleanarchitecture.domain.model.weather.WeatherModel;
import com.weathercock.android_cleanarchitecture.domain.repository.WeatherRepository;

import org.modelmapper.ModelMapper;

import io.realm.Realm;
import retrofit2.Retrofit;
import rx.Observable;

/**
 * Weather Repository Implement
 * <p>
 * Created by tamura_k on 2016/05/27.
 */
public class WeatherRepositoryImpl implements WeatherRepository {

    private Context context;

    public WeatherRepositoryImpl(Context context) {
        this.context = context;
    }

    @Override
    public Observable<WeatherModel> fetch(int cityId) {
        Retrofit retrofit = WeatherRetrofit.getInstance();
        WeatherApi weatherApi = retrofit.create(WeatherApi.class);
        ModelMapper modelMapper = new ModelMapper();
        return weatherApi.get(cityId)
                .doOnNext(weatherEntity -> {
                    if (exist(cityId)) delete(cityId);
                    weatherEntity.setCityId(cityId);
                    insert(weatherEntity);
                })
                .map(weatherEntity -> modelMapper.map(weatherEntity, WeatherModel.class));
    }

    @Override
    public WeatherModel find(int cityId) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            WeatherDao weatherDao = new WeatherDao(realm);
            WeatherEntity weatherEntity = weatherDao.find(cityId);

            if (weatherEntity != null) {
                ModelMapper modelMapper = new ModelMapper();
                return modelMapper.map(weatherEntity, WeatherModel.class);
            } else {
                return null;
            }
        } finally {
            if (realm != null) realm.close();
        }
    }

    private boolean exist(int cityId) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            WeatherDao weatherDao = new WeatherDao(realm);
            return weatherDao.exist(cityId);
        } finally {
            if (realm != null) realm.close();
        }
    }

    private void insert(WeatherEntity weatherEntity) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            realm.beginTransaction();
            WeatherDao weatherDao = new WeatherDao(realm);
            weatherDao.insert(weatherEntity);
            realm.commitTransaction();
        } finally {
            if (realm != null) realm.close();
        }
    }

    private void delete(int cityId) {
        Realm realm = null;
        try {
            realm = Realm.getDefaultInstance();
            realm.beginTransaction();
            WeatherDao weatherDao = new WeatherDao(realm);
            weatherDao.delete(cityId);
            realm.commitTransaction();
        } finally {
            if (realm != null) realm.close();
        }
    }

}

File: data/repository/src/main/java/com/kazakago/blueprint/data/repository/hierarchy/weather/WeatherRepositoryImpl.kt
package com.kazakago.blueprint.data.repository.hierarchy.weather

import android.content.Context
import com.kazakago.blueprint.data.api.hierarchy.weather.WeatherApi
import com.kazakago.blueprint.data.database.entity.weather.DescriptionEntity
import com.kazakago.blueprint.data.database.entity.weather.ForecastEntity
import com.kazakago.blueprint.data.database.entity.weather.LocationEntity
import com.kazakago.blueprint.data.database.entity.weather.WeatherEntity
import com.kazakago.blueprint.data.database.global.AppDatabase
import com.kazakago.blueprint.data.repository.global.dispatcher.CacheFlowDispatcher
import com.kazakago.blueprint.data.repository.mapper.weather.*
import com.kazakago.blueprint.domain.model.global.state.State
import com.kazakago.blueprint.domain.model.global.state.mapContent
import com.kazakago.blueprint.domain.model.hierarchy.city.CityId
import com.kazakago.blueprint.domain.model.hierarchy.weather.Weather
import com.kazakago.blueprint.domain.repository.hierarchy.weather.WeatherRepository
import com.os.operando.guild.kt.Quartet
import com.os.operando.guild.kt.to
import kotlinx.coroutines.flow.Flow

internal class WeatherRepositoryImpl(context: Context) : WeatherRepository {

    private val weatherApi = WeatherApi(context)
    private val weatherDao = AppDatabase.create(context).weatherDao()
    private val weatherResponseMapper = WeatherResponseMapper(LocationResponseMapper(), DescriptionResponseMapper(), ForecastResponseMapper())
    private val weatherEntityMapper = WeatherEntityMapper(LocationEntityMapper(), DescriptionEntityMapper(), ForecastEntityMapper())

    override fun subscribe(cityId: CityId): Flow<State<Weather>> {
        return getDispatcher(cityId)
            .subscribe(needRefresh = { it.first.isExpired() })
            .mapContent {
                weatherEntityMapper.map(it.first, it.second, it.third, it.fourth)
            }
    }

    override suspend fun request(cityId: CityId) {
        return getDispatcher(cityId).request()
    }

    private fun getDispatcher(cityId: CityId): CacheFlowDispatcher<Quartet<WeatherEntity, LocationEntity, DescriptionEntity, List<ForecastEntity>>> {
        return CacheFlowDispatcher(
            stateId = WeatherEntity::class.qualifiedName + cityId.value,
            loadEntity = { loadEntity(cityId) },
            fetchOrigin = {
                val fetched = weatherApi.fetch(cityId.value)
                val model = weatherResponseMapper.map(fetched, cityId)
                weatherEntityMapper.reverse(model)
            },
            saveEntity = { saveEntity(it.first, it.second, it.third, it.fourth) }
        )
    }

    private suspend fun loadEntity(cityId: CityId): Quartet<WeatherEntity, LocationEntity, DescriptionEntity, List<ForecastEntity>>? {
        val weather = weatherDao.findWeather(cityId.value)
        val location = weatherDao.findLocation(cityId.value)
        val description = weatherDao.findDescription(cityId.value)
        val forecasts = weatherDao.findForecasts(cityId.value)
        return if (weather != null && location != null && description != null && forecasts != null) {
            return weather to location to description to forecasts
        } else {
            null
        }
    }

    private suspend fun saveEntity(weather: WeatherEntity, location: LocationEntity, description: DescriptionEntity, forecasts: List<ForecastEntity>) {
        weatherDao.insertWeather(weather)
        weatherDao.insertLocation(location)
        weatherDao.insertDescription(description)
        weatherDao.insertForecasts(forecasts)
    }
}
